docs/site/Discovering-models-from-relational-databases.md

```diff
@@ -0,0 +1,49 @@
+---
+lang: en
+title: 'Discovering models from relational databases'
+keywords: LoopBack 4.0, LoopBack-Next
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Discovering-models-from-relational-databases.html
+---
+
+## Synopsis
+
+LoopBack makes it simple to create models from an existing relational database.
+This process is called _discovery_ and is supported by the following connectors:
+
+- [Cassandra](Cassandra-connector.html)
+- [MySQL](MySQL-connector.html)
+- [Oracle](Oracle-connector.html)
+- [PostgreSQL](PostgreSQL-connector.html)
+- [SQL Server](SQL-Server-connector.html)
+- [IBM DB2](DB2-connector.html)
+- [IBM DashDB](DashDB.html)
+- [IBM DB2 for z/OS](DB2-for-z-OS.html)
+- [SAP HANA](https://www.npmjs.org/package/loopback-connector-saphana) - Not
+  officially supported; see [Community connectors](Community-connectors.html).
+
+## Overview
+
+Models can be discovered from a supported datasource by running the
+`lb4 discover` command.
+
+**The LoopBack project must be built and contain the built datasource files in
+`PROJECT_DIR/dist/src/datasources.js`**
+
+### Options
+
+`--dataSource`: Put a valid datasource name here to skip the datasource prompt
+
+`--views`: Choose whether to discover views. Default is true
+
+`--all`: Skips the model prompt and discovers all of them
+
+`--outDir`: Specify the directory into which the `model.model.ts` files will be
+placed
+
+`--schema`: Specify the schema which the datasource will find the models to
+discover
+
+`--schemaDefs`: The path to a `.json` file containing a model definition, like
```

Comment:
Can we find a better name for this option, what that will convey the fact that we are loading models from JSON files instead of the database?

Few ideas to consider:

- `--import`
- `--import-definition`
- `--from-definition`
- `--from-file`
- `--from-lb3-models`

Also please make mention that the JSON file can contain either a single model definition object or an array with multiple definitions.

Reply:
On the second thought, I am not sure if it's a good idea to mix `lb4 discover` with importing models from JSON definitions. I find it little bit confusing as a user, plus it makes the implementation more complex because of all those `if (importing from file) return` statements at the top of many generator steps.

Have you considered moving this functionality into `lb4 model` generator? I think it will give us CLI that feels pretty natural too.

```sh
lb4 model --import definition.json
# or perhaps
lb4 model --from-definition definition.json
```

Reply:
Yes it's dramatically more intuitive, I'll move it to the model generator.  

Slightly related:  The `.json` loading was used to give me a way of testing the model discovery, which I still have a bit of trouble doing thoroughly.  A couple of solutions:

The memory datasource could be given `discoverModelDefinitions` and `discoverSchema`, but that might be out of the scope of this PR.  

Docker could also be used in the unit tests, but that definitely out of scope of this PR

Reply:
@bajtos After looking around the code a bit I'm proposing keeping the --schemaDefs option.  And that I submit a seperate PR for adding the `--import` option to the model generator.

The current `discover --schemaDefs path` accepts an array of objects, which mimics how the `discover` command discovers multiple models.  I also need a way to mock data for the integration test.

Reply:
Let's discuss testing and user experience independently.

For testing, I think you should mock `discoverModelDefinitions` and/or `discoverSchema` in such way that your tests sees the model schema you are interested in. I'll try to find time to create a code snippet showing how to mock those two methods.

For user experience, I still prefer importing via `lb4 model` over `--schemaDefs`. I am fine with leaving this option out of this initial pull request, as long as `--schemaDefs` is removed too.

Reply:
I am thinking about the following solution for testing:

- Modify the datasource class used by your tests (`mem.datasource.js.txt`) to override discovery methods provided by `juggler.DataSource`.
- In these overridden methods, load schema from `test-schema-defs.json` and return the relevant parts depending on which discovery method was called.

A mock-up to illustrate what I mean:

```ts
export class MemDataSource extends juggler.DataSource {
  async discoverSchema(name, options) {
    await ensureSchemaDefsLoaded();
    const schema = this.schemas.find(s => s.name === name);
    if (schema) return schema;
    throw new Error(`Schema not found: ${name}`);
  }

  private schemas;
  private async ensureSchemaDefsLoaded() {
    if (schemas) return;
    const schemaFile = path.resolve(__dirname, 'test-schema-defs.json');
    _schemas = JSON.parse(await fs.readFile(schemaFile, 'utf-8'));
  }
}    
```

Reply:
I'm pretty embarrassed I didn't think of implementing my own DataSource.  The testing is now much easier.    

https://github.com/strongloop/loopback-next/blob/5445c58c09760df39f5d010a136979f36c57b217/packages/cli/test/fixtures/discover/mem.datasource.js.txt